### PR TITLE
Faster Determinant Solving by adding Triangular Matrix Check #18607

### PR DESCRIPTION
**Pull request by sudoWin**
#### References to other Issues or PRs

https://github.com/sympy/sympy/issues/18607 by @sylee957

#### Brief description of what is fixed or changed
Check for Triangular Matrix added in Determinant Solver Function
Earlier, for all matrices, either 'bareiss' or the specified
method was used for computing the determinant.
Now it is first checked whether it is a triangular matrix
and if it is, it allows determinant computation by just
product of diagonal elements

Similary _charpoly has been modified to check Triangular Matrix.
If it is a Trinagular Matrix then, diagonal elememts are also eigen values and
they are used to compute the characteristic polynomial, which is faster
than Berkowitz Algorithm

#### Other comments

Also updated the docstrings, added some more tests for
_charpoly and _det

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/matrices/determinant.py [hunk 1]
```diff
@@ -396,6 +401,13 @@ def _charpoly(M, x='lambda', simplify=_simplify):
 
     if not M.is_square:
         raise NonSquareMatrixError()
+    if M.is_lower or M.is_upper:
+        diagonal_elements = M.diagonal()
+        x = _uniquely_named_symbol(x, diagonal_elements)
+        m = 1
+        for i in diagonal_elements:
+            m = m * (x - simplify(i))
+        return PurePoly(m,x)
```
Comment (sylee957) [inline sympy/matrices/determinant.py]:
```suggestion
        return PurePoly(m, x)
```

File: sympy/matrices/determinant.py [hunk 2]
```diff
@@ -557,7 +573,12 @@ def _det(M, method="bareiss", iszerofunc=None):
     n = M.rows
 
     if n == M.cols: # square check is done in individual method functions
-        if n == 0:
+        if M.is_upper or M.is_lower:
+            m = 1
+            for i in range(n):
+                m = m * M[i,i]
```
Comment (sylee957) [inline sympy/matrices/determinant.py]:
```suggestion
                m = m * M[i, i]
```

**Comment by sudoWin** [comment]
Added spaces as suggested in the above review by @sylee957

File: sympy/matrices/determinant.py [hunk 3]
```diff
@@ -486,6 +498,10 @@ def _det(M, method="bareiss", iszerofunc=None):
         specified method is ignored. Otherwise, it defaults to
         ``'bareiss'``.
 
+        Also,if the matrix is an upper or a lower triangular matrix, determinant
```
Comment (sylee957) [inline sympy/matrices/determinant.py]:
```suggestion
        Also, if the matrix is an upper or a lower triangular matrix, determinant
```

**Pull request merged**